osbenchmark/synthetic_data_generator/synthetic_data_generator.py

```diff
@@ -6,98 +6,148 @@
-            dask_client,
-            sdg_config,
-            raw_mappings,
-            mapping_config
-        )
-
-        write_record_and_publish_summary_to_console(sdg_config, total_time_to_generate_dataset, generated_dataset_details)
-
-def orchestrate_data_generation(cfg):
-    logger = logging.getLogger(__name__)
-    sdg_config = create_sdg_config_from_args(cfg)
-    test_document_requested = cfg.opts("synthetic_data_generator", "test_document")
-
-    # TODO: Rename custom config
-    custom_config = load_config(sdg_config.custom_config_path) if sdg_config.custom_config_path else {}
-
-    # TODO: Move client creation deeper so that cancellation is easier on the command line
-    workers = custom_config.get("settings", {}).get("workers", os.cpu_count())
-    dask_client = Client(n_workers=workers, threads_per_worker=1)  # We keep it to 1 thread because generating random data is CPU intensive
-    logger.info("Number of workers to use: %s", workers)
-
-    console.println(f"[NOTE] Dashboard link to monitor processes and task streams: {dask_client.dashboard_link}")
-    console.println("[NOTE] For users who are running generation on a virtual machine, consider tunneling to localhost to view dashboard.")
-    console.println("")
-
-    if host_has_available_disk_storage(sdg_config):
-        if use_custom_synthetic_data_generator(sdg_config):
-            orchestrate_data_generation_for_custom_synthetic_data_generator(test_document_requested, sdg_config, custom_config, dask_client)
-        elif use_mappings_synthetic_data_generator(sdg_config):
-            orchestrate_data_generation_for_mapping_synthetic_data_generator(test_document_requested, sdg_config, dask_client)
-    else:
-        logger.error("User wants to generate [%s]GB but current host does not have enough available disk storage.", sdg_config.total_size_gb)
-        console.println(f"User wants to generate [{sdg_config.total_size_gb}]GB but current host does not have enough available disk storage")
-        sys.exit(1)
+from osbenchmark.synthetic_data_generator.types import GB_TO_BYTES
+from osbenchmark.synthetic_data_generator.helpers import write_chunk, get_generation_settings, setup_custom_tqdm_formatting, check_for_exsiting_files
+from osbenchmark.synthetic_data_generator.strategies import DataGenerationStrategy
+from osbenchmark.synthetic_data_generator.types import SyntheticDataGeneratorMetadata
+
+class SyntheticDataGenerator:
+    def __init__(self, sdg_metadata: SyntheticDataGeneratorMetadata, sdg_config: dict, strategy: DataGenerationStrategy) -> None:
+        self.sdg_metadata = sdg_metadata
+        self.sdg_config = sdg_config
+        self.strategy = strategy
+
+        self.logger = logging.getLogger(__name__)
+
+    def generate_seeds_for_workers(self, regenerate=False):
+        # This adds latency so might consider deprecating this
```

Comment:
Perhaps this can be pregenerated and saved, if it does not necessarily need to be derived from the worker ID.

Reply:
I can open a subsequent PR that addresses this as this one is getting long. I debated on what the seed should be generated off of for a while and settled on Worker ID + Time to ensure that the new values are unique. This is because if Dask somehow detects the same seed is used before, it'll use the previous cached response, which would introduce duplicate documents. We can further discuss this in a separate issue.